### DISCUSSION TITLE
classfield vs method

### DISCUSSION THREAD

**Discussion by lokba** [woowacourse-study/2022-code-review-study#38] [Lotto] [answered]
1번과 2번의 차이점은 뭘까요? 또한, console의 답또한 예측해보세요.
```javascript
class Crew {
  name;
  nickname;
  constructor() {}

  //1번
  makeName = (name) => {
    this.name = name;
  };


  //2번
  makeNickname(nickname) {
    this.nickname = nickname;
  }
}

const lokba = new Crew();
const marco = new Crew();

//문제
console.log(lokba.makeNickname === marco.makeNickname);
console.log(lokba.makeName === marco.makeName);

```

아래 코드에서 어떤 방식을 선호하시나요? 그리고 그 이유는 무엇인가요?
```javascript
class Machine {
  name;
  constructor() {
    //1번
    this.$lokba.addEventListener("click", this.lokbaClickEvent.bind(this));

    //2번
    this.$kkojae.addEventListener("click", this.kkojaeClickEvent);
  }

  //1번
  lokbaClickEvent() {
    this.name = "lokba";
  }

  //2반
  kkojaeClickEvent = () => {
    this.name = "kkojae";
  };
}
```

**Comment by byhhh2** [comment]
흑흑 .. 누가.. 이렇게 어려운 문제를..
답은..!

```js
//문제
console.log(lokba.makeNickname === marco.makeNickname); // true
console.log(lokba.makeName === marco.makeName); // false
```
ㅎ..ㅎ..

일단 쥐어짠 지식의 풀이는

method : prototype에 정의되어 있어서 모든 인스턴스에 공유됨
classfield에 정의한 화살표함수 : prototype에 정의되지 않음 

`makeNickname`은 공유되기 때문에 lokba나 marco의 `makeNickname`가 같은거죠 !

---

두번째는 ~!

저는 
```js
this.$kkojae.addEventListener("click", (e) => this.kkojaeClickEvent(e)); // 3번
```

3번으로 씁니다!

이유는 

1. `bind()`해주기 귀찮음.
2. 화살표함수로 클래스의 메서드를 만들면 안됨

**Comment by moonheekim0118** [comment]
### 첫번째 질문 
문제는 위의 무비가 풀어줬으니까 패스할게요 (ㅋㅋㅋ)
저 예전에 기술면접봤을 때 이 질문 받았었어요.  
제가 그 때 아래와 같이 정리했었네요~
- 클래스에 정의된 화살표 함수는 멤버변수로 할당이 되고, 일반 메서드는 프로토타입 프로퍼티로 할당이 된다.
- 클래스 메서드는 프로토타입에 정의되어 있으며, 모든 인스턴스에서 공유된다. -> 만약 10000개의 인스턴스를 만들더라도 자바스크립트 엔진이 히든 클래스를 생성하여 모든 인스턴스마다 메서드를 들고 다니지 않게 도와준다. 하지만, 클래스 멤버변수에 화살표 함수를 할당한다면, 10000개의 인스턴스를 만들 경우 10000개의 함수를 만들게 된다. 프로토타입의 프로퍼티가 아니기 때문이다.
- 리액트 컴포넌트는 '상속'이 아니라 '합성'을 추구하기 때문에 문제가 되지는 않아서 사용하고 있는 것이다. (리액트에서는 클래스 내부 화살표 함수 정의)

[히든클래스 관련 글 추천](https://meetup.toast.com/posts/78)

### 두번째 질문

화살표 함수를 정의하는건 선호하지 않으니까 1번이요 ..~ 저도 무비처럼 아래와 같이 쓰네요
```javascript
this.$kkojae.addEventListener("click", (e) => this.kkojaeClickEvent(e)); 
```

그런데 제가 로이님 리뷰에서 아래와 같은 피드백을 본적이 있어요 👀...  로이님은 클래스필드 선언을 나쁘지 않게 보시나봅니다..

> 메소드 선언 방식에 의하면 prototype에 원본메소드가 남아있는 채로, 실제로 호출되는건 새로 만들어진 bound 메소드입니다. 사용하지 않을 불필요한 메소드가 남아있다는 뜻이죠. 반면 class field를 이용하면 prototype에는 남아있지 않고, 인스턴스 자신이 사용할 함수만 들고 있게 됩니다.

**Reply by wonsss** [reply]
```js
this.$kkojae.addEventListener("click", (e) => this.kkojaeClickEvent(e)); 
```
이렇게 쓰면, 해당 이벤트리스너를 remove할 때 어떤 방법을 사용하시나요?

**Reply by moonheekim0118** [reply]
위 사용법의 단점이 바로 remove 를 못한다는 것이죠.. 예리하시네요~

**Comment by wonsss** [comment, marked as answer]
![image](https://user-images.githubusercontent.com/59413128/158161006-24cbefde-2db0-46db-b9cd-c8ebfa00c04e.png)

![image](https://user-images.githubusercontent.com/59413128/158161157-81728c4f-fef8-4bfa-a414-7bfde046dbe1.png)

일반 함수 방식의 메서드는 각 인스턴스에서 해당 메서드를 호출했을 때, [[Prototype]]을 통해 original 메서드 하나를 공유해서 쓰고 있네요.
Crew.prototype.MakeNickname 으로도 호출가능하죠!

반면 화살표 함수 방식의 메서드는 [[Prototype]] 체계에 속하지 않고,  각 인스턴스마다 생성되는 멤버변수 같은 거군요.  따라서 화살표 함수 방식의 메서드는 인스턴스마다 서로 다른거구요.
여기서는 lobka.MakeName 처럼 인스턴스를 통해서만 호출할 수 있구요.

Babel에서 트랜스파일링하여 더 확실하게 살펴보시죠.

```js
class Crew {
  name;
  nickname;
  constructor() {}

  //1번
  makeName = (name) => {
    this.name = name;
  };


  //2번
  makeNickname(nickname) {
    this.nickname = nickname;
  }
}
```
```js
// 트랜스파일한 코드
"use strict";

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class Crew {
  constructor() {
    _defineProperty(this, "name", void 0);

    _defineProperty(this, "nickname", void 0);

    _defineProperty(this, "makeName", name => {
      this.name = name;
    });
  } //1번


  //2번
  makeNickname(nickname) {
    this.nickname = nickname;
  }

}
```
보면 특이하게도, makeName은 name, nickname과 같이 프로퍼티로서 멤버변수에 할당되고 있었네요.
makeNickname은 여전히 일반 메서드 형태구요.

결과적으로,  메서드로 화살표함수를 사용한다면, 인스턴스를 생성할 때마다 메서드로서 새로운 객체를 생성하여 성능 저하의 문제가 있을 것 같네요.  

더 자세한 내용은 다음 링크에서 살펴보시죠(저도 나중에 봐야겠습니다..흑흑)
https://medium.com/@charpeni/arrow-functions-in-class-properties-might-not-be-as-great-as-we-think-3b3551c440b1

---

두 번째로, 이벤트 리스너의 바인딩 방식 관련해서 제 생각은 다음과 같습니다. 
저는 1번을 선호하구요. 

추가적으로 이벤트 리스너가 binding되는 방식(bind, 화살표함수) 등의 방식에 대해 고민했는데, 결국 나중에 이벤트 리스너를 제거하는 것이 중요하다는 결론에 이르렀습니다.

따라서 this바인딩된 이벤트 리스너(콜백함수)를 제거할 수 있는 좋은 방법은 무엇인지 생각해봤는데, 별도의 내용이긴 하지만 같이 알아두면 좋을 것 같아요.

그 방법은 this 바인딩된 이벤트리스너를 constructor 안에서 멤버변수로 할당한 후, 이벤트를 추가하거나 제거하는 경우 이벤트리스너 인자 자리에 멤버변수를 넣어주는 것입니다. 아래처럼요.
```js
class Marco {
  constructor() {
     this.myListener = this.myListener.bind(this);
  }
  $myElement.addEventListener('click', this.myListener);
  $myElement.removeEventListener('click', this.myListener);
}
```
이 방법에 대해서 어떻게 생각하시나요?

**Reply by wonsss** [reply]
https://engineering.linecorp.com/ko/blog/v8-hidden-class/

**Comment by kkojae91** [comment]
## 첫 번째
첫 번째 질문 쉽지 않네요.. 한참 생각하고 내려가서 다른 크루들이 써주신것 보고 이해했어요!
프로토타입 공부를 해야겠다 생각이드네요..

## 두 번째
저는 보통 `bind(this)`를 선호했는데, 이번에 같이 하게된 크루가 전부 화살표 함수를 사용하더라구요! 그래서 따라 하긴 했지만,
저는 bind(this)를 해주는게 더 좋은것 같아요!

**Comment by wonsss** [comment]
bind하면 해당 메서드는 어디에 선언되나요? bound라는 영역이 따로 있나요?
